### PR TITLE
(packaging)Update paths to AIO layout

### PR DESCRIPTION
**Pull request by shrug**
Update to ezbake 0.3.7 and associated path changes to reflect the layout specified in https://github.com/puppetlabs/puppet-specifications/blob/master/file_paths.md.

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -126,7 +126,7 @@
              :ezbake {:dependencies ^:replace [[puppetlabs/puppetdb ~pdb-version]
                                                [org.clojure/tools.nrepl "0.2.3"]]
                       :name "puppetdb"
-                      :plugins [[puppetlabs/lein-ezbake "0.2.9"
+                      :plugins [[puppetlabs/lein-ezbake "0.3.7"
```
Comment (haus) [inline project.clj]: With this change it should be safe to pull lines 141 and 142 (assuming we want PE and FOSS on the same ezbake version.

**Comment by haus** [comment]
@shrug this all looks good to me. I have that 1 comment, but it shouldn't block merging.

File: project.clj (hunk 1 above)
Comment (shrug) [inline project.clj]: Good call, updated to remove the redundant plugin dependency

**Comment by kbarber** [comment]
@shrug @haus this one is still blocked on the testing work. We're very warm on this front, so don't worry. Thank you so much for the PR - its awesome. Right now we need to:

a) fix the master Puppet 4 testing job so it passes
b) merge that job with the existing all-in-one job (and remove the old one)
c) update the PR testing so it tests against Puppet 4 (right now it tests against Puppet 3 only)

I'm going to dive in first thing this week, this won't continue to be blocked for long. /cc @johnduarte

**Comment by kbarber** [comment]
Oh there is this general error also during puppetdb ssl-setup run:

```
el5-64-1 executed in 5.86 seconds
Exited: 1
#<Beaker::Host::CommandFailure: Host 'el5-64-1' exited with 1 running:
 cd /root/puppetdb-3.0.0.SNAPSHOT.2015.05.29T1510; bash install.sh service
Last 10 lines of output were:
	PEM files in /etc/puppetlabs/puppetdb/ssl are missing, we will move them into place for you
	Copying files: /etc/puppet/ssl/certs/ca.pem, /etc/puppet/ssl/private_keys/el5-64-1.us-west-2.compute.internal.pem and /etc/puppet/ssl/certs/el5-64-1.us-west-2.compute.internal.pem to /etc/puppetlabs/puppetdb/ssl
	chown: `pe-puppetdb:pe-puppetdb': invalid user>
```

Looks like the conditionals around choosing the user here aren't quite right.

**Comment by kbarber** [comment]
Actually, we need puppetdb ssl-setup fixed so its AIO/non-AIO agnostic. I think this is the pathing issues @haus has alluded to before in https://tickets.puppetlabs.com/browse/PDB-1455. This is what I believe is blocking the tests for master.

**Comment by kbarber** [comment]
The ssl-setup user issue is because of this: 
https://github.com/puppetlabs/puppetdb/blob/master/resources/ext/cli/ssl-setup#L231-L237

Its presuming its a PE setup, when its not.

**Comment by kbarber** [comment]
So I think this is either going to need a template for ssl-setup, so that there can be a build-time decision for "if PE or not PE" ... or we get fancy some other way. Maybe interrogate the pe-puppetdb user, or something other forensic check. Build time feels better.

**Comment by haus** [comment]
@kbarber I agree about template/build-time being preferable over runtime.

**Comment by kbarber** [comment]
Looks like this has to go in for our tests to go green for master now :-). So yeah, slightly messed up the ordering in my head.

Okay, I've got a job geared up to test this patch with now, so we don't have to merge blindly:

https://jenkins.puppetlabs.com/view/PuppetDB/view/master%20-%20Ezbake/job/platform_puppetdb_intn-sys-aio-kbtest/

I think @shrug ... if we can get the ssl-setup fix done first, we can then fire up that job and see if this patch is safe based on that. That way we're not interrupting the existing pipeline.

**Comment by kbarber** [comment]
I'm testing out this patch with a fix for ssl-setup on top: https://github.com/kbarber/puppetdb/commit/b1b2286362212de5b18348b8e3de61506ba25d16

Job is here, but it might require more fiddling: https://jenkins.puppetlabs.com/view/PuppetDB/view/Other%20Builds/job/platform_puppetdb_intn-sys_source-kbtest/2/

**Comment by kbarber** [comment]
The test ^^ failed, on el7 at least it looks like we're seeing incorrect pathing when using the module. In acceptance/helper.rb, we use the class puppetdb::server and puppetdb::server::database_ini quite frequently, these need a hand with a confdir parameter ... at least until we can fix the module to use the AIO defaults. Its a bit chicken and egg, we don't consume the latest version of the module in git, only on the forge, and we can't release a 5.0 specifically. This will work around it for now however ... until we circle back round to the module.

**Comment by kbarber** [comment]
I've provided a patch in my branch that might fix the issue, new job is here:

https://jenkins.puppetlabs.com/view/PuppetDB/view/Other%20Builds/job/platform_puppetdb_intn-sys_source-kbtest/3/

My branch is here: https://github.com/kbarber/puppetdb/tree/aio_paths

I'll be logging off soon, feel free to pick this work up, otherwise I'll continue tomorrow.

**Comment by shrug** [comment]
I got wrapped up in the PE merges today so I did not get a chance to get back to this today, hopefully we can sync back up tomorrow

**Comment by kbarber** [comment]
Looks like last night we had the disk on sector fill up and those tests bombed, when I tried to rerun it it was toast: https://tickets.puppetlabs.com/browse/QENG-2512. Waiting on QENG to solve it so I can re-run.

**Comment by kbarber** [comment]
Latest tests show a terminus install prob: https://jenkins.puppetlabs.com/view/PuppetDB/view/Other%20Builds/job/platform_puppetdb_intn-sys_source-kbtest/6/BEAKER_CONFIG=ec2-west-el7-64mda-el7-64a,BEAKER_OPTIONS=postgres,BEAKER_TYPE=aio,label=beaker-ec2/consoleFull

**Comment by kbarber** [comment]
Looks like the latest failure is due to source based installation of the termini not installing things into the correct location:

```
# EZ_VERBOSE=true bash install.sh termini
+ prefix=/usr
+ initdir=/etc/init.d
+ unitdir=/usr/lib/systemd/system
+ datadir=/usr/share
+ real_name=puppetdb
+ projdatadir=/usr/share/puppetdb
+ confdir=/etc
+ projconfdir=/etc/puppetlabs/puppetdb
+ rundir=/var/run/puppetlabs/puppetdb
+ bindir=/opt/puppetlabs/server/apps/puppetdb/bin
+ uxbindir=/opt/puppetlabs/bin
+ symbindir=/opt/puppetlabs/server/bin
+ localstatedir=/var
+ app_prefix=/opt/puppetlabs/server/apps/puppetdb
+ app_data=/opt/puppetlabs/server/data/puppetdb
+ app_logdir=/var/log/puppetlabs/puppetdb
+ '[' -n true ']'
+ set +x
#-------------------------------------------------#
The following variables are set:

_=/bin/env
EZ_VERBOSE=true
G_BROKEN_FILENAMES=1
HISTCONTROL=ignoredups
HISTSIZE=1000
HOME=/root
HOSTNAME=el6-64-1
LANG=en_US.UTF-8
LESSOPEN=||/usr/bin/lesspipe.sh %s
LOGNAME=root
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;05;37;41:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.tbz=01;31:*.tbz2=01;31:*.bz=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.axa=01;36:*.oga=01;36:*.spx=01;36:*.xspf=01;36:
MAIL=/var/spool/mail/root
PATH=/opt/puppetlabs/bin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin
PWD=/root/puppetdb-3.0.0.SNAPSHOT.2015.06.02T0749
SELINUX_LEVEL_REQUESTED=
SELINUX_ROLE_REQUESTED=
SELINUX_USE_CURRENT_RANGE=
SHELL=/bin/bash
SHLVL=2
SSH_AUTH_SOCK=/tmp/ssh-DUERH11066/agent.11066
SSH_CLIENT=88.96.178.182 61063 22
SSH_CONNECTION=88.96.178.182 61063 10.128.103.200 22
SSH_TTY=/dev/pts/0
TERM=xterm
USER=root

End of variable print.
#-------------------------------------------------#
+ dispatch termini
+ local task=termini
+ shift
+ '[' -z termini ']'
+ '[' '' = -h ']'
+ task termini
+ local task=termini
+ shift
+ debug_echo 'Running task termini ...'
+ '[' -n true ']'
+ echo Running task termini ...
Running task termini ...
+ eval task_termini
++ task_termini
+++ /opt/puppet/bin/ruby -rrbconfig -e 'puts RbConfig::CONFIG['\''sitelibdir'\'']'
+++ ruby -rrbconfig -e 'puts RbConfig::CONFIG['\''sitelibdir'\'']'
++ local rubylibdir=/usr/lib/ruby/site_ruby/1.8
++ install -Dm 0644 puppet/indirector/node/puppetdb.rb /usr/lib/ruby/site_ruby/1.8/puppet/indirector/node/puppetdb.rb
++ install -Dm 0644 puppet/indirector/facts/puppetdb_apply.rb /usr/lib/ruby/site_ruby/1.8/puppet/indirector/facts/puppetdb_apply.rb
++ install -Dm 0644 puppet/indirector/facts/puppetdb.rb /usr/lib/ruby/site_ruby/1.8/puppet/indirector/facts/puppetdb.rb
++ install -Dm 0644 puppet/indirector/catalog/puppetdb.rb /usr/lib/ruby/site_ruby/1.8/puppet/indirector/catalog/puppetdb.rb
++ install -Dm 0644 puppet/indirector/resource/puppetdb.rb /usr/lib/ruby/site_ruby/1.8/puppet/indirector/resource/puppetdb.rb
++ install -Dm 0644 puppet/util/puppetdb/command.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb/command.rb
++ install -Dm 0644 puppet/util/puppetdb/command_names.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb/command_names.rb
++ install -Dm 0644 puppet/util/puppetdb/config.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb/config.rb
++ install -Dm 0644 puppet/util/puppetdb/blacklist.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb/blacklist.rb
++ install -Dm 0644 puppet/util/puppetdb/http.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb/http.rb
++ install -Dm 0644 puppet/util/puppetdb/char_encoding.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb/char_encoding.rb
++ install -Dm 0644 puppet/util/puppetdb.rb /usr/lib/ruby/site_ruby/1.8/puppet/util/puppetdb.rb
++ install -Dm 0644 puppet/application/storeconfigs.rb /usr/lib/ruby/site_ruby/1.8/puppet/application/storeconfigs.rb
++ install -Dm 0644 puppet/reports/puppetdb.rb /usr/lib/ruby/site_ruby/1.8/puppet/reports/puppetdb.rb
++ install -Dm 0644 puppet/face/node/deactivate.rb /usr/lib/ruby/site_ruby/1.8/puppet/face/node/deactivate.rb
++ install -Dm 0644 puppet/face/node/status.rb /usr/lib/ruby/site_ruby/1.8/puppet/face/node/status.rb
++ install -Dm 0644 puppet/face/storeconfigs.rb /usr/lib/ruby/site_ruby/1.8/puppet/face/storeconfigs.rb
```

Could be that an ezbake change is required.

**Comment by kbarber** [comment]
I've got a patch for the ruby termini autodetection stuff here: https://github.com/puppetlabs/ezbake/pull/248 ... we'll need that reviewed & merged so we can work on a snapshot. For this PR work to continue.

**Comment by kbarber** [comment]
Ezbake fix is in, but until we get a release - I've switched over to using ezbake_dev_build which pulls down and build ezbake from master on its own. Pretty trivial temporary patch: https://github.com/kbarber/puppetdb/commit/5c2325c5318a67b95f14a527e0e817359f344e76

Tests are running again here: https://jenkins.puppetlabs.com/view/PuppetDB/view/Other%20Builds/job/platform_puppetdb_intn-sys_source-kbtest/8/

**Comment by kbarber** [comment]
Okay, So I've gotten the pathing wrong for AIO, its meant to use vendordir, not sitelibdir or whatever:

https://github.com/kbarber/ezbake/commit/5b430d084af85fe2f3b99c029686b1a744761f20

Notice 'vendordir' here? Thats the only thing that works, and I see this reflected in our stable branch.

What I'm confused about is this:

https://github.com/puppetlabs/ezbake/blob/master/resources/puppetlabs/lein-ezbake/template/foss/ext/redhat/ezbake.spec.erb#L28

I can't see how redhat would work here at all, it looks like there is no AIO pathing assumption for redhat here, I'm guessing nobody else even uses this ... I presume all the work is in PE only.

Debian seems fine: https://github.com/puppetlabs/ezbake/blob/master/resources/puppetlabs/lein-ezbake/template/foss/ext/debian/rules.erb#L7

I think that redhat spec file needs change before packages would even build, or I could be wrong.

Anyway ... testing again:

https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_source-kbtest/11/

**Comment by haus** [comment]
@kbarber yea the foss spec looks wrong. puppetdb is the first project to ship a foss terminus off master of ezbake, which is the why.

**Comment by kbarber** [comment]
@haus okay, I've fixed the rspec thing to match the PE version in my pending ezbake branch: https://github.com/kbarber/ezbake/tree/ticket/master/pdb-1455 ...

**Comment by kbarber** [comment]
Okay two subtle acceptance tests failures last run, 1 was related to the to_json call we use in puppet 4 not working as intended in puppet server. I've rolled this back to to_pson for now, and left a big red mark in the comments next to it. Not sure what the right thing is to do here.

The other problem looked like a testing failure only, it was how lines are reported - seems in puppet 4 it expresses the last line as the line that changed, in puppet 3 it uses the first line? Anyway, this test has to change to reflect that the notify is now on line 1 in a puppet world :-).
https://github.com/puppetlabs/puppetdb/blob/master/acceptance/tests/reports/report_storage.rb#L83

Tests running again:

https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_source-kbtest/13/

**Comment by kbarber** [comment]
Last two bugs squashed, one more test failure to do with slight variations in how puppet 4 works - nothing major, commit pushed & another test going:

https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_source-kbtest/14/

**Comment by kbarber** [comment]
Looks like we have a green build:

https://jenkins.puppetlabs.com/view/PuppetDB/view/Other%20Builds/job/platform_puppetdb_intn-sys_source-kbtest/15/

Working branch is here:

https://github.com/kbarber/puppetdb/commits/aio_paths

I've raised an ezbake PR @haus and @shrug, it's over here:

https://github.com/puppetlabs/ezbake/pull/250

The only pending thing I think I need to work out is what to do about `to_pson`. Previously we had switched to_json, and I reverted that in my branch to make tests pass. But without a clever hash traversal conversion to convert the data to utf8 _before_ the to_json, I'm not sure what the right move is here. to_pson still continues to work however - and it's a very quick move to get us green and merged. I'm going to sleep on this.

https://github.com/kbarber/puppetdb/commit/26ca4fc1d7da2447b103303bb25ba62ec44830f7
https://gist.github.com/kbarber/28fd0990141c0dbdd3aa
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_source-kbtest/12/BEAKER_CONFIG=ec2-west-el7-64mda-el7-64a,BEAKER_OPTIONS=postgres,BEAKER_TYPE=aio,label=beaker-ec2/consoleFull

**Comment by kbarber** [comment]
Come to think of it, the puppetserver emits json via the catalog back to the agent, the same catalog we are ourselves are trying to convert to json ^^. I think there is a solution for this in Puppet core somewhere.

**Comment by kbarber** [comment]
I've raised a new cleaned up PR here: https://github.com/puppetlabs/puppetdb/pull/1436

**Pull request closed without merging**